### PR TITLE
C++ API

### PR DESCRIPTION
**Pull request by jackyko1991**
This is a fork of https://github.com/tensorflow/tensorflow/pull/24963

1.  To build a DLL for TensorFlow C++ API, you can now build
  `//tensorflow:tensorflow_cc.dll`
2. You can also get a compressed package for both (tensorflow.dll, tensorflow.lib) and C API headers by building
  `//tensorflow/tools/lib_package:libtensorflow_cc`

### CODE REVIEW

**Comment by meteorcloudy** [comment]
@jackyko1991 #24963 is now merged. Can you rebase your branch?

There're something wrong during code transformation when pushing to github at https://github.com/tensorflow/tensorflow/blob/3e21fe5faedab3a8258d344c8ad1cec2612a8aa8/tensorflow/BUILD#L521, I'm sending a fix to add the missing line endings.

**Comment by jackyko1991** [comment]
@meteorcloudy I have rebased my fork to match your fixes.

When I come across with the C++ api, I discovered that it depends a lot on the core libraries which some of the headers maybe moved or renamed. 

For example, `tensorflow::Status` is one of the mostly used class in TF core, its [header](https://github.com/tensorflow/tensorflow/blob/3f23f4ddeabbdc0704444d84c158bd6c348a9f10/tensorflow/core/lib/core/status.h) includes a non-existing `#include "tensorflow/core/lib/core/error_codes.pb.h"`, which I believe is renamed as [tensorflow/core/lib/core/errors.h](https://github.com/tensorflow/tensorflow/blob/3f23f4ddeabbdc0704444d84c158bd6c348a9f10/tensorflow/core/lib/core/errors.h)

strangely enough this kind of file renaming still pass CI tests.

Another issue is that ops now appears to be named as "kernels"

In the [official document of c++ api](https://www.tensorflow.org/guide/extend/cc), it has the line `using namespace tensorflow::ops;`, it was defined by this [header](https://github.com/tensorflow/tensorflow/blob/3f23f4ddeabbdc0704444d84c158bd6c348a9f10/tensorflow/cc/ops/standard_ops.h). But now it is relocated at [core kernels](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/kernels). This will ruin most of the old c++ codes without a clear clarification on header changes.

I am working at exporting core headers together with `
 //tensorflow/tools/lib_package:libtensorflow_cc`, but library linking and third party dependency may still be problematic, especially for eigen and other lower level headers like absl.

Can you come with a better idea to solve it?

**Comment by meteorcloudy** [comment]
@jackyko1991 Sorry, I'm not sure the correct solution here because I'm not quite familiar with actual implementation.

@mrry Can you help answer the questions?

**Comment by mrry** [comment]
@gunan owns the ongoing process of modularizing the TF codebase and providing stable APIs to other libraries. I believe he has plans in motion, so I'll defer to him on how a large change to the build files would fit with those plans.

File: tensorflow/core/BUILD [hunk 1]
```diff
@@ -5224,3 +5224,358 @@ alias(
...
+        "lib/strings/proto_text_util.h",
+        "lib/strings/scanner.h",
+        "lib/strings/str_util.h",
+        "lib/strings/strcat.h",
+        "lib/strings/stringprintf.h",
+    ],
+    visibility = ["//visibility:public"],
+)
+
+# filegroup(
```
Comment (meteorcloudy) [inline tensorflow/core/BUILD]: Remove this?

**Review by meteorcloudy — CHANGES_REQUESTED** [review]
Although I'm not familiar with the actual headers, the structure of this PR makes sense to me. 
But it would be even better if we can add a few tests for this.

File: tensorflow/core/BUILD (hunk 1 above)
Comment (jackyko1991) [reply tensorflow/core/BUILD]: This has been fixed

**Comment by chsigg** [comment]
The devil is probably in the details, but shouldn't transitive_hdrs be able to collect those header files?

**Comment by jackyko1991** [comment]
I have tried transitive_hdrs at the beginning but I would like the headers
to locate at original folders structures. Directly use transitive_hdrs
seems only flatten all headers in one folder layer?

Christian Sigg <notifications@github.com> 於 2019年3月6日週三 上午5:43寫道：

**Comment by chsigg** [comment]
No, it should not flatten the directory tree. You can use a bazel genrule to write the headers with the correct tree structure.

**Comment by jackyko1991** [comment]
@chsigg do you mean use some genrule like the following:
```bazel
genrule(
    name = "tensorflow_cc_headers",
    srcs = [
        "//tensorflow/core:headers"
    ],
    outs = tf_genrule_cmd_cc_headers,
    cmd = "cp $(locations //tensorflow/core:headers) > $@",
)
```

I am new to bazel and using some hard skills to copy the header. It will be nice if you can provide relavent example on copying file in folder structure.

**Comment by jackyko1991** [comment]
https://github.com/tensorflow/tensorflow/issues/22047#issuecomment-421452033 suggested an approach to export necessary symbols via `tf_exported_symbols_msvc.lds`.

Can anyone give a help on this? My temporary solution is to link all the generated libs against the executable but still get some of them missing.

**Comment by guikarist** [comment]
> [#22047 (comment)](https://github.com/tensorflow/tensorflow/issues/22047#issuecomment-421452033) suggested an approach to export necessary symbols via `tf_exported_symbols_msvc.lds`.
> 
> Can anyone give a help on this? My temporary solution is to link all the generated libs against the executable but still get some of them missing.

It is just a last resort. See [guide in my script](https://github.com/guikarist/tensorflow-windows-build-script#prerequisites)(after *"If you need to build C++ API,"*). Then my script apply [a patch](https://github.com/guikarist/tensorflow-windows-build-script/blob/master/patches/cpp_symbol.1.12.0.patch) to Tensorflow source before building.

**Comment by chsigg** [comment]
> @chsigg do you mean use some genrule like the following:
> ...

Would something like the install_headers target work?

https://github.com/tensorflow/tensorflow/blob/08d767919de9c5698d18c3de1c17c3e6a0dfa961/tensorflow/BUILD#L664

**Comment by bygreencn** [comment]
> [#22047 (comment)](https://github.com/tensorflow/tensorflow/issues/22047#issuecomment-421452033) suggested an approach to export necessary symbols via `tf_exported_symbols_msvc.lds`.
> 
> Can anyone give a help on this? My temporary solution is to link all the generated libs against the executable but still get some of them missing.

cmake compile mode give create_def_file.py to give a method, but bazel compile mode had not do it and current tensorflow_cc has not exported the correct cc symbols. Could you please continue to check it make tensorflow__cc export the correct symbols?

**Comment by bygreencn** [comment]
> > [#22047 (comment)](https://github.com/tensorflow/tensorflow/issues/22047#issuecomment-421452033) suggested an approach to export necessary symbols via `tf_exported_symbols_msvc.lds`.
> > Can anyone give a help on this? My temporary solution is to link all the generated libs against the executable but still get some of them missing.
> 
> cmake compile mode give create_def_file.py to give a method, but bazel compile mode had not do it and current tensorflow_cc has not exported the correct cc symbols based on your patch with v1.13.1. Could you please continue to check it make tensorflow__cc export the correct symbols?

image
I means that it still has more c++ api with TF_*  prefix and CPP project will feedback lnk2001 error.

**Comment by guikarist** [comment]
Any news?

**Comment by rthadur** [comment]
@jackyko1991 can you please resolve conflicts

**Comment by rthadur** [comment]
@jackyko1991 can you please resolve conflicts ?

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**